configuration.md
# Server configuration

Warrant requires certain configuration variables to be set via either a `warrant.yaml` config file or via environment variables. There is a set of common variables as well as datastore-specific configuration.

## Common variables

| Variable | Description | Required? | Default | YAML | ENV VAR |
| -------- | ----------- | --------- | ------- | ---- | ------- |
| `port` | Port where the server runs. | no | 8000 | `port: VALUE` | `WARRANT_PORT=VALUE` |
| `logLevel` | Log level (e.g. Debug, Info etc.) for the server. Warrant uses zerolog, valid log levels are defined [here](https://github.com/rs/zerolog#leveled-logging). | no | 0 | `logLevel: VALUE` | `WARRANT_LOGLEVEL=VALUE` |
| `enableAccessLog` | Determines whether the built-in request logger is enabled or not. | no | true | `enableAccessLog: VALUE` | `WARRANT_ENABLEACCESSLOG=VALUE` |
| `autoMigrate` | If set to `true`, the server will apply datastore migrations before starting up. | no | false | `autoMigrate: VALUE` | `WARRANT_AUTOMIGRATE=VALUE` |
| `check.concurrency` | The default concurrency setting for access checks. | no | 4 | `concurrency: VALUE` | `WARRANT_CHECK_CONCURRENCY=VALUE` |
| `check.maxConcurrency` | The max concurrency setting for access checks. | no | 1000 | `maxConcurrency: VALUE` | `WARRANT_CHECK_MAXCONCURRENCY=VALUE` |
| `check.timeout` | Access check global timeout. | no | 1m | `timeout: VALUE` | `WARRANT_CHECK_TIMEOUT=VALUE` |

## Warrant Server Authentication
Warrant supports two types of authentication: API key and JWT authentication tokens.

### API Key Authentication
By default, you must configure an API key that Warrant will use to authenticate all requests. You should follow standard security practices for generating and storing your API key.

| Variable | Description | Required? | Default | YAML | ENV VAR |
| -------- | ----------- | --------- | ------- | ---- | ------- |
| `authentication.apiKey` | The unique API key that all clients must pass to the server via the `Authorization: ApiKey VALUE` header | yes | - | `authentication:`<br>&emsp;`apiKey: VALUE` | `WARRANT_AUTHENTICATION_APIKEY=VALUE` |

### 3rd-party Auth Provider Token Authentication
You can optionally configure Warrant to allow access check requests made to the `/v2/authorize` endpoint using JWT authentication tokens generated by your application or a 3rd-party authentication provider (e.g. Auth0, Firebase, etc). You can also configure the claims in the JWT token that specify the `userId` and `tenantId` of the user being authenticated. These claims will be used to automatically populate the subject and context for the access check(s) being made, so any requests using JWTs will be scoped to the user and tenant specified in the token.

| Variable | Description | Required? | Default | YAML | ENV VAR |
| -------- | ----------- | --------- | ------- | ---- | ------- |
| `authentication.provider` | The authentication provider used to generate the auth tokens. | yes | - | `authentication:`<br>&emsp;`provider: VALUE` | `WARRANT_AUTHENTICATION_PROVIDER=VALUE` |
| `authentication.publicKey` | The signing certificate used to sign the auth token. Currently only RSA signed tokens are supported. | yes | - | `authentication:`<br>&emsp;`publicKey: VALUE` | `WARRANT_AUTHENTICATION_PUBLICKEY=VALUE` |
| `authentication.userIdClaim` | The claim containing the user id of the user being authenticated. | no | sub | `authentication:`<br>&emsp;`userIdClaim: VALUE` | `WARRANT_AUTHENTICATION_USERIDCLAIM=VALUE` |
| `authentication.tenantIdClaim` | The claim containing the tenant id of the user being authenticated. | no | - | `authentication:`<br>&emsp;`tenantIdClaim: VALUE` | `WARRANT_AUTHENTICATION_TENANTIDCLAIM=VALUE` |

If you are using Firebase as your authentication provider, the public key value is optional.

## Set up datastore

Warrant is a stateful service that runs with an accompanying `datastore`. Currently, `MySQL`, `PostgreSQL` and `SQLite` (file and in-memory) are supported. Refer to these guides to set up your desired database(s):

- [MySQL](/migrations/datastore/mysql/README.md)
- [PostgreSQL](/migrations/datastore/postgres/README.md)
- [SQLite](/migrations/datastore/sqlite/README.md)

Here is an example of a full server config using `mysql` for the datastore:

### Sample `warrant.yaml` config (place file in same dir as server binary)

```yaml
port: 8000
logLevel: 1
enableAccessLog: true
autoMigrate: true
check:
    concurrency: 4
    maxConcurrency: 1000
    timeout: 1m
authentication:
    apiKey: your_api_key
datastore:
  mysql:
    username: replace_with_username
    password: replace_with_password
    hostname: replace_with_hostname
    database: warrant
```

### Sample `warrant.yaml` config with JWT authentication config
```yaml
port: 8000
logLevel: 1
enableAccessLog: true
autoMigrate: true
check:
    concurrency: 4
    maxConcurrency: 1000
    timeout: 1m
authentication:
    apiKey: your_api_key
    provider: auth0
    publicKey: |
        -----BEGIN CERTIFICATE-----
        your_public_signing_key
        -----END CERTIFICATE-----
    userIdClaim: sub
    tenantIdClaim: aud
datastore:
  mysql:
    username: replace_with_username
    password: replace_with_password
    hostname: replace_with_hostname
    database: warrant
```

### Sample environment variables config

```shell
export WARRANT_PORT=8000
export WARRANT_LOGLEVEL=1
export WARRANT_ENABLEACCESSLOG=true
export WARRANT_AUTOMIGRATE=true
export WARRANT_CHECK_CONCURRENCY=4
export WARRANT_CHECK_MAXCONCURRENCY=1000
export WARRANT_CHECK_TIMEOUT="1m"
export WARRANT_AUTHENTICATION_APIKEY="replace_with_api_key"
export WARRANT_DATASTORE_MYSQL_USERNAME="replace_with_username"
export WARRANT_DATASTORE_MYSQL_PASSWORD="replace_with_password"
export WARRANT_DATASTORE_MYSQL_HOSTNAME="replace_with_hostname"
export WARRANT_DATASTORE_MYSQL_DATABASE="warrant"
```


pkg/config/config.go
// Copyright 2023 Forerunner Labs, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package config

import (
	"fmt"
	"os"
	"reflect"
	"strings"
	"time"

	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
	"github.com/rs/zerolog/pkgerrors"
	"github.com/spf13/viper"
)

const (
	DefaultMySQLDatastoreMigrationSource    = "github://warrant-dev/warrant/migrations/datastore/mysql"
	DefaultPostgresDatastoreMigrationSource = "github://warrant-dev/warrant/migrations/datastore/postgres"
	DefaultSQLiteDatastoreMigrationSource   = "github://warrant-dev/warrant/migrations/datastore/sqlite"
	DefaultAuthenticationUserIdClaim        = "sub"
	PrefixWarrant                           = "warrant"
	ConfigFileName                          = "warrant.yaml"
)

type Config interface {
	GetPort() int
	GetLogLevel() int8
	GetEnableAccessLog() bool
	GetAutoMigrate() bool
	GetDatastore() *DatastoreConfig
}

type WarrantConfig struct {
	Port            int              `mapstructure:"port"`
	LogLevel        int8             `mapstructure:"logLevel"`
	EnableAccessLog bool             `mapstructure:"enableAccessLog"`
	AutoMigrate     bool             `mapstructure:"autoMigrate"`
	Datastore       *DatastoreConfig `mapstructure:"datastore"`
	Authentication  *AuthConfig      `mapstructure:"authentication"`
	Check           *CheckConfig     `mapstructure:"check"`
}

func (warrantConfig WarrantConfig) GetPort() int {
	return warrantConfig.Port
}

func (warrantConfig WarrantConfig) GetLogLevel() int8 {
	return warrantConfig.LogLevel
}

func (warrantConfig WarrantConfig) GetEnableAccessLog() bool {
	return warrantConfig.EnableAccessLog
}

func (warrantConfig WarrantConfig) GetAutoMigrate() bool {
	return warrantConfig.AutoMigrate
}

func (warrantConfig WarrantConfig) GetDatastore() *DatastoreConfig {
	return warrantConfig.Datastore
}

func (warrantConfig WarrantConfig) GetAuthentication() *AuthConfig {
	return warrantConfig.Authentication
}

func (warrantConfig WarrantConfig) GetCheck() *CheckConfig {
	return warrantConfig.Check
}

type DatastoreConfig struct {
	MySQL    *MySQLConfig    `mapstructure:"mysql"`
	Postgres *PostgresConfig `mapstructure:"postgres"`
	SQLite   *SQLiteConfig   `mapstructure:"sqlite"`
}

type MySQLConfig struct {
	Username                 string        `mapstructure:"username"`
	Password                 string        `mapstructure:"password"`
	Hostname                 string        `mapstructure:"hostname"`
	Database                 string        `mapstructure:"database"`
	MigrationSource          string        `mapstructure:"migrationSource"`
	MaxIdleConnections       int           `mapstructure:"maxIdleConnections"`
	ConnMaxIdleTime          time.Duration `mapstructure:"connMaxIdleTime"`
	MaxOpenConnections       int           `mapstructure:"maxOpenConnections"`
	ConnMaxLifetime          time.Duration `mapstructure:"connMaxLifetime"`
	ReaderHostname           string        `mapstructure:"readerHostname"`
	ReaderMaxIdleConnections int           `mapstructure:"readerMaxIdleConnections"`
	ReaderMaxOpenConnections int           `mapstructure:"readerMaxOpenConnections"`
	DSN                      string        `mapstructure:"dsn"`
	ReaderDSN                string        `mapstructure:"readerDsn"`
}

type PostgresConfig struct {
	Username                 string        `mapstructure:"username"`
	Password                 string        `mapstructure:"password"`
	Hostname                 string        `mapstructure:"hostname"`
	Database                 string        `mapstructure:"database"`
	SSLMode                  string        `mapstructure:"sslmode"`
	MigrationSource          string        `mapstructure:"migrationSource"`
	MaxIdleConnections       int           `mapstructure:"maxIdleConnections"`
	ConnMaxIdleTime          time.Duration `mapstructure:"connMaxIdleTime"`
	MaxOpenConnections       int           `mapstructure:"maxOpenConnections"`
	ConnMaxLifetime          time.Duration `mapstructure:"connMaxLifetime"`
	ReaderHostname           string        `mapstructure:"readerHostname"`
	ReaderMaxIdleConnections int           `mapstructure:"readerMaxIdleConnections"`
	ReaderMaxOpenConnections int           `mapstructure:"readerMaxOpenConnections"`
}

type SQLiteConfig struct {
	Database           string        `mapstructure:"database"`
	InMemory           bool          `mapstructure:"inMemory"`
	MigrationSource    string        `mapstructure:"migrationSource"`
	MaxIdleConnections int           `mapstructure:"maxIdleConnections"`
	ConnMaxIdleTime    time.Duration `mapstructure:"connMaxIdleTime"`
	MaxOpenConnections int           `mapstructure:"maxOpenConnections"`
	ConnMaxLifetime    time.Duration `mapstructure:"connMaxLifetime"`
}

type AuthConfig struct {
	ApiKey   string              `mapstructure:"apiKey"`
	Provider *AuthProviderConfig `mapstructure:"providers"`
}

type AuthProviderConfig struct {
	Name          string `mapstructure:"name"`
	PublicKey     string `mapstructure:"publicKey"`
	UserIdClaim   string `mapstructure:"userIdClaim"`
	TenantIdClaim string `mapstructure:"tenantIdClaim"`
}

type CheckConfig struct {
	Concurrency    int           `mapstructure:"concurrency"`
	MaxConcurrency int           `mapstructure:"maxConcurrency"`
	Timeout        time.Duration `mapstructure:"timeout"`
}

func NewConfig() WarrantConfig {
	viper.SetConfigFile(ConfigFileName)
	viper.SetDefault("port", 8000)
	viper.SetDefault("logLevel", zerolog.DebugLevel)
	viper.SetDefault("enableAccessLog", true)
	viper.SetDefault("autoMigrate", false)
	viper.SetDefault("datastore.mysql.connMaxIdleTime", 4*time.Hour)
	viper.SetDefault("datastore.mysql.connMaxLifetime", 6*time.Hour)
	viper.SetDefault("datastore.mysql.migrationSource", DefaultMySQLDatastoreMigrationSource)
	viper.SetDefault("datastore.postgres.connMaxIdleTime", 4*time.Hour)
	viper.SetDefault("datastore.postgres.connMaxLifetime", 6*time.Hour)
	viper.SetDefault("datastore.postgres.migrationSource", DefaultPostgresDatastoreMigrationSource)
	viper.SetDefault("datastore.sqlite.connMaxIdleTime", 4*time.Hour)
	viper.SetDefault("datastore.sqlite.connMaxLifetime", 6*time.Hour)
	viper.SetDefault("datastore.sqlite.migrationSource", DefaultSQLiteDatastoreMigrationSource)
	viper.SetDefault("authentication.providers.userIdClaim", DefaultAuthenticationUserIdClaim)
	viper.SetDefault("check.concurrency", 4)
	viper.SetDefault("check.maxConcurrency", 1000)
	viper.SetDefault("check.timeout", 1*time.Minute)

	// If config file exists, use it
	_, err := os.ReadFile(ConfigFileName)
	if err == nil {
		if err := viper.ReadInConfig(); err != nil {
			log.Fatal().Err(err).Msg("init: error while reading warrant.yaml. Shutting down.")
		}
	} else {
		if os.IsNotExist(err) {
			log.Info().Msg("init: could not find warrant.yaml. Attempting to use environment variables.")
		} else {
			log.Fatal().Err(err).Msg("init: error while reading warrant.yaml. Shutting down.")
		}
	}

	var config WarrantConfig
	// If available, use env vars for config
	for _, fieldName := range getFlattenedStructFields(reflect.TypeOf(config)) {
		envKey := strings.ToUpper(fmt.Sprintf("%s_%s", PrefixWarrant, strings.ReplaceAll(fieldName, ".", "_")))
		envVar := os.Getenv(envKey)
		if envVar != "" {
			viper.Set(fieldName, envVar)
		}
	}

	if err := viper.Unmarshal(&config); err != nil {
		log.Fatal().Err(err).Msg("init: error while creating config. Shutting down.")
	}

	// Configure logger
	zerolog.TimeFieldFormat = zerolog.TimeFormatUnix
	zerolog.DurationFieldUnit = time.Millisecond
	zerolog.ErrorStackMarshaler = pkgerrors.MarshalStack
	zerolog.SetGlobalLevel(zerolog.Level(config.GetLogLevel()))
	if zerolog.GlobalLevel() == zerolog.DebugLevel {
		log.Logger = log.Output(zerolog.ConsoleWriter{Out: os.Stderr})
	}

	if config.GetAuthentication() == nil || config.GetAuthentication().ApiKey == "" {
		log.Fatal().Msg("init: must provide an API key to authenticate incoming requests to Warrant.")
	}

	return config
}

func getFlattenedStructFields(t reflect.Type) []string {
	return getFlattenedStructFieldsHelper(t, []string{})
}

func getFlattenedStructFieldsHelper(t reflect.Type, prefixes []string) []string {
	unwrappedT := t
	if t.Kind() == reflect.Pointer {
		unwrappedT = t.Elem()
	}

	flattenedFields := make([]string, 0)
	for i := 0; i < unwrappedT.NumField(); i++ {
		field := unwrappedT.Field(i)
		fieldName := field.Tag.Get("mapstructure")
		switch field.Type.Kind() {
		case reflect.Struct, reflect.Pointer:
			flattenedFields = append(flattenedFields, getFlattenedStructFieldsHelper(field.Type, append(prefixes, fieldName))...)
		default:
			flattenedField := fieldName
			if len(prefixes) > 0 {
				flattenedField = fmt.Sprintf("%s.%s", strings.Join(prefixes, "."), fieldName)
			}
			flattenedFields = append(flattenedFields, flattenedField)
		}
	}

	return flattenedFields
}
